---
title: "Assignment 5 - DESeq2"
output: html_document
---


For more detailed instructions, please read the assignment page in the course
textbook. Please note that for this assignment, many of the next steps depend on
the previous working and for this reason, some of the tests are also dependent
on the previous function working correctly. Run your tests one at a time to
start, and check that each step is working as expected before moving on.

Use this block to source your `main.R`.

```{r setup, message=FALSE, warning=FALSE}
source('main.R')
```

## 1. Reading and subsetting the data from verse_counts.tsv and sample_metadata.csv

To begin, we will subset the counts matrix to only contain a subset of samples
found in the original data. We have provided you with the *full* counts matrix
of samples and a metadata file containing relevant experimental variables for
each sample. These files are named 'verse_counts.tsv' and 'sample_metadata.csv',
respectively. Define a function that reads from these two files to construct a
counts matrix and matching sample dataframe containing only samples belonging to
the vP0 and vAd timepoints. This should correspond to a total of 4 samples
(vP0_1, vP0_2, vAd_1, vAd_2). Store this counts matrix and the sample dataframe
in a SummarizedExperiments object. Ensure that the name of your counts matrix
stored as an assay in the SummarizedExperiment object is `counts`.

Also check that the order of your samples in your colData is the same as they
are ordered in your counts matrix. Before storing your sample dataframe as
colData, ensure you set the factor levels so that vP0 is the reference level for
the column that will be used as input to the design matrix, in our case,
`timepoint`. By default,the first reference level will be considered the
reference. Your colData should have columns named `samplename` and `timepoint` as
seen in the reference report.

```{r, warning=FALSE, message=FALSE, echo=FALSE}
#Read in verse_counts.tsv and sample_metadata.csv and subset to the listed samples
#Store these dataframes in a SummarizedExperiment object and display the summary here
se <- make_se('data/verse_counts.tsv', 'data/sample_metadata.csv', c('vP0', 'vAd'))
se
```

Display the colData from your SummarizedExperiment object here:

```{r, echo=FALSE}
#Display the sample data from your SE object here to make sure the order is right 
# Assuming 'se' is your SummarizedExperiment object
sample_data <- colData(se)

# Display only 'samplename' and 'timepoint' columns
sample_data[, c('samplename', 'timepoint')]
```

## 2. [Running DESeq2](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis)

DESeq2 has a built-in function that directly uses a SummarizedExperiment object.
Define a function that takes your SummarizedExperiment and a design formula and
runs DESeq2 with standard parameters. In our case, we wish to determine what
genes are differentially expressed between our samples based on timepoint and
our model will simply be `~ timepoint` and we will use `vP0` as our reference
level. 

Return a *list* containing both the results from DESeq2 as a dataframe as well
as the DESeqDataSet object updated with various statistics in the course of the
analysis (referred to as the dds in the vignette)

```{r, run deseq2, echo=FALSE}
#Run your DESeq2 function here which returns a list of the dds and the results
#Assign the results and the dds from the returned list to their own variables 
#for ease of use
design_formula <- ~ timepoint

# Call the function with the correct design formula
results_list <- return_deseq_res(se, design = design_formula)
```

## 3. Annotating results to construct a labeled volcano plot

Using the results generated by DESeq2, convert it to a `tibble` and add one
additional column that denotes whether a gene is either: 1. Upregulated and
significant at a given padj threshold, 2.Downregulated and significant at a
given padj threshold, or 3. Not significant at a given padj threshold. __Name
this column `volc_plot_status`. Ensure the values for these labels are UP, DOWN,
and NS, respectively. __ Have your function take the DESeq2 results as well as a
user-defined threshold for padj and set it at .10.

```{r, label results tibble, echo=FALSE}
#Label the results with the column as specified above
#Display the summary of the tibble
# Call the function to label the DESeq2 results and create the tibble
results_df <- as.data.frame(results)

# Create a labeled results tibble as shown in a previous code example
labeled_results <- label_res(results_df, padj_threshold = 0.10)
labeled_results

```

### Captions for figures

```{r code block for descriptive captions}

caption_4 <- 'Histogram of raw pvalues obtained from DE analysis (vPO vs. VAd)'
caption_5 <- 'Histogram of Log2FoldChanges for DE Genes (vPO vs. VAd)'
caption_6 <- 'Plot of Log10 (normalized counts) for top ten DE genes'
caption_7 <- 'Volcano plot of DESeq2 differential expression results (vPO vs. VAd)'
caption_9 <- 'fgsea results for Hallmark MSigDB gene se'

```

## 4. Diagnostic plot of the raw p-values for all genes

Using the same results tibble from the previous function, plot a histogram of
the raw p-values for all of the genes discovered in the experiment. Avoid using
the base ggplot2 theme and make sure to label any appropriate axes and
components of the plot.

```{r, raw pval histogram, echo=FALSE, warning=FALSE, fig.cap = caption_4}
#Histogram of raw p-values for all discovered genes
# Call the plot_pvals function to create a histogram
pval_histogram <- plot_pvals(labeled_results)

# Display the histogram
print(pval_histogram)
```

## 5. Plotting the LogFoldChanges for differentially expressed genes

Using the results tibble, make a histogram of the Log2FoldChange values for all
significant differentially expressed genes at a user-defined padj threshold and
set this threshold to .10. Avoid using the base ggplot2 theme and make sure to
label any appropriate axes and components of the plot.

```{r, log2fc DE genes histogram, echo=FALSE, fig.cap = caption_5}
#Histogram of log2FC values for significant DE genes with padj (FDR) < .10
# Call the plot_log2fc function to create a histogram
log2fc_histogram <- plot_log2fc(labeled_results, padj_threshold = 0.10)

# Display the histogram
print(log2fc_histogram)
```

## 6. Plotting the normalized counts of differentially expressed genes

Extract out the DESeq2 normalized counts (by size factors) for the top ten
differentially expressed genes as ranked by ascending padj. Make a scatter / 
jitter plot of the counts and label each by their samplenames. 

```{r, plot of normalized counts, echo=FALSE, fig.cap = caption_6}
#Make a scatter/jitter plot of the norm counts for the top ten DE genes ranked
#by padj
# Call the scatter_norm_counts function to create a scatter/jitter plot
norm_counts_plot <- scatter_norm_counts(labeled_results, dds, num_genes = 10)

# Display the plot
print(norm_counts_plot)

```

## 7. Volcano Plot to visualize differential expression results

Using the status column you generated in part 3, make a volcano plot of the
DESeq2 results and color the points appropriately. 

```{r, volcano plot, echo=FALSE, warning=FALSE, fig.cap = caption_7}
#Volcano Plot (scatter plot with log2FC on x-axis and -log10(padj) on y-axis)
#Color points (genes) by status defined in part 3
volcano_plot <- plot_volcano(labeled_results)

# Display the volcano plot
print(volcano_plot)
```

## 8. Running fgsea

This section will be split into two separate functions that will generated a
ranked named list of genes and their log2FoldChange values and perform
`fgsea()`, respectively. Please use the provided id2gene.txt file you containing
matching ensembl ids and their corresponding gene symbol to add a column to your
labeled results for the corresponding gene symbol. Then, use this new column and
generate a named vector containing your gene symbols and their corresponding
log2FC value ordered by log2FC descending. Make sure to use the gene symbols in
place of their ensembl id in order to match the identifiers used in the GMT gene
sets file.

```{r, generate ranked list of log2FC}
#first load in the id2gene.txt appropriately
#add a new column in your labeled results that matches IDs to symbols
#generate a named vector of symbols and log2FC values in descending order
rnk_list <- make_ranked_log2fc(labeled_results, 'data/id2gene.txt')

```

```{r, this is the "top" of our ranked list}
#use head to display the top of the list
head(rnk_list)
```

```{r, this is the "bottom" of our ranked list}
#use tail to display the bottom of the list
tail(rnk_list)
```

Perform a GSEA using a ranked list log2FoldChange values for all genes
discovered in the DESeq2 results against the C2 Canonical Pathways gene set
collection. As recommended, please set the minSize and maxSize parameters to 15
and 500, respectively. Before running fgsea, you will have to load the GMT
pathways file in an appropriate format for `fgsea`. Otherwise, run fgsea using
default parameters and the `fgsea()` function.

```{r, actually run fgsea, warning=FALSE, message=FALSE, echo=FALSE}
#Run fgsea using a ranked list of descending log2FC against the C2 canonical
#pathways gene set
#Set minsize to 15 and maxsize to 500, leave the other parameters as defaults
fgsea_results <- run_fgsea('data/m2.cp.v2023.1.Mm.symbols.gmt', rnk_list, 15, 500)

```

```{r, fgsea results, echo=FALSE}
#Display the results of the fgsea in a tibble sorted by padj (ascending)
# Sort the results by padj in ascending order
sorted_results <- fgsea_results %>% arrange(padj)
print(sorted_results)
```
## 9. Plotting the top ten positive NES and top ten negative NES pathways

To make a quick visualization of these GSEA results, make a barchart for the
pathways with the top ten most positive and top ten most negative NES scores.
Color the bars by whether they have a positive or negative NES.

```{r, fgsea results plot, echo=FALSE, fig.cap = caption_9}
#Plot the top ten pathways with both positive and negative NES (20 total)
#Color the pathways by the sign of their NES (positive or negative)
fgsea_plot <- top_pathways(fgsea_results, 10)
fgsea_plot
```

