---
title: "Assignment 2 Report"
author: "Taylor Falk"
date: "11/17/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

**Please read this brief introduction to R Markdown, and complete the second
half of the document yourself.**

# R Markdown
R Markdown, `.Rmd` files, are a type of markdown files that can execute R code
in an R environment. Markdown is an open source language used for easily
creating documents such as web pages and PDFs. Markdown and R Markdown can be
used to procedurally generate documents and reports, which have a wide number
of use-cases across bioinformatics. Not only can you save time by creating code
to automatically draft documents, you can also include detailed instructions
and information alongside your code.

## knitr
`knitr` is simply the package R uses to convert the R Markdown into the
appropriate output (HTML, PDF, or Word). While there is a deep level of
customization available, most users can comfortably create beautiful documents
with the default parameters. In order to compile your markdown, in RStudio you
can select the `Knit` button below the open file button.

---

## Markdown Basics
The raw markdown, which you may or may not be looking at, is styled very
basically. The largest titles start with one octothorpe `#`. The more `#`'s
added, the smaller a title or header becomes.  

# Biggest  
## 2nd Biggest  
### so on  
#### and so forth  
##### until you have  
###### six octothorpes  
####### then nothing happens.

Importantly, two spaces (`  `) can be included after a line of text to start a
new line.

Markdown can be styled a few different ways:  
`_italics_` -> _italics_  
`**bold**` -> **bold**  
`` `monospace` `` -> `monospace`, good for code  

A good cheat sheet is the [Markdown Guide Cheat
Sheet](https://www.markdownguide.org/cheat-sheet)

---

## R Code in Markdown

The really vital part of R Markdown is the inclusion of code snippets you can
place into the document, which will run and print out to users. In this snippet
I create a function that says hello and that text is printed inside the
document.
```{r}
say_hi <- function() {
  return("Hello world.")
}
print(say_hi())
```

The syntax of the **snippet* is simple, with three back ticks ` ``` `
indicating the start and end of the snippet, and a lower case `r` in curly
braces `{}` to indicate what language is used: `{r}` indicates R, `{python}`
indicates Python. R Markdown does not support every language, but there are use
cases where one document running both R and Python code can prove extremely
useful and concise. There is also a button in RStudio to insert a snippet, it
is a small green `+C` in the top right of the editor.

```{python}
import datetime
print(f'Wishing you a wonderful hello from Python at {datetime.datetime.now()}')
```

I can also test my code in RStudio without having to knit the entire document
by pressing the green "Play" arrow on the snippet I am interested in.

Finally, R Markdown scripts can `source()` other `.R` scripts and use them in
their execution environment. This mean I can define a function in my `main.R`
script and use it in my R Markdown snippets. Note that the converse is true,
using `knitr::purl()`, and the code established in an R Markdown can be
sourced, and _tested_, using an normal `.R` script.

Also note that the environment is carried from one snippet to another, so if I
define a function or load a file in one snippet I can use those objects in
later snippets. 

```{r}
message <- 'The last example...for now.'
```

```{r}
print(paste(message, 'But there\'s always room for more.'))
```


# Assignment
Using what you know from this document and the `main.R` script, use the
functions from `main.R` to draw a boxplot of  expression levels. Use this R
Markdown document to describe your code as you are using it in this document,
as done above. Take time to enhance your ggplot function from the `main.R`
script with high quality visual aesthetics. Do not use default ggplot colors.

<!-- TODO: Add a figure -->
```{r}
# use source to load your functions from main.R into this document.
source("main.R")
library(tidyverse)
```
## Loading CSV with `load_expression()`

Importing data into R effectively is crucial. For our tasks, we prefer the tibble format, a modern representation of data frames in R. Load the data into a tibble as we have done in
prior assignments. 

```{r load_expression}

# Load the expression data
result_tib <- load_expression("data/example_intensity_data_subset.csv")
result_tib # Display the initial rows of the data
```
## Filtering Rows with `filter_15()`

To effectively work with our slightly vast dataset, it's essential to filter out
rows that don't meet certain criteria. The `filter_15()` function does precisely
this by retaining rows where at least 15% of the values surpass a log2(15)
expression levelâ€”equivalent to approximately 3.9. Notably, the function returns
only the probe IDs, providing a summarized view of the dataset that meets the
criteria.

While one might initially consider using traditional loops for row-wise
operations, such methods can be inefficient for large datasets. Leveraging
functions from the Tidyverse, particularly from the `dplyr` package, offers a
more efficient and readable approach. Functions like `filter()` and `mutate()`
allow for swift data transformations, enhancing both clarity and performance.


```{r filter_15}
filtered_probes <- c(filter_15(result_tib)$probe)
print(paste0("Original number of probes in unfiltered data: ", nrow(result_tib)))
print(paste0("Probes remaining in filtered dataset: ", length(filtered_probes))) 

#tibble of probes that survive the filter
filter_15(result_tib)
```

## Converting Affy IDs to HGNC Names: `affy_to_hgnc()`

Gene data often involves various identifiers, and converting between these can
be crucial for data integration. The function `affy_to_hgnc()` serves this
purpose by linking Affymetrix probe IDs to HGNC gene IDs using the `biomaRt`
package. However, while `biomaRt` offers a direct connection to Ensembl, a
comprehensive genomic database, its reliance on external APIs can lead to
occasional connectivity issues. If errors arise during usage, they're often due
to these external connections rather than the code itself.

When implementing, remember: 

- Use the `ENSEMBL_MART_ENSEMBL` biomart and the `hsapiens_gene_ensembl` dataset. 

- Fetch the attributes `"affy_hg_u133_plus_2"` and `"hgnc_symbol"`. 

- Despite the function aiming to return a tibble, `biomaRt`'s `getBM()` only
deals with data.frames. Utilize `dplyr::pull()` to transform a tibble to a
character vector, and `dplyr::as_tibble()` for converting a data frame back to a
tibble.
```{r affy_to_hgnc}
# Converting a sample affy ID to HGNC
filter_15(result_tib) %>% affy_to_hgnc() -> gene_names
gene_names
```
## Refining Data with reduce_data()
Our data preparation's ultimate goal is to streamline the dataset, focusing on relevant genes to make visualization more efficient with ggplot from the tidyverse package. The reduce_data() function serves this purpose by integrating various inputs, such as:

The original expression data.
Probe IDs linked to HGNC symbols.
Lists of 'good' and 'bad' gene names.
The function operates as follows:

- Matches probe IDs with HGNC symbols using the base function match().

- Introduces the new data at the appropriate position using tibble::add_column().

- Segregates the genes into two categories, 'good' and 'bad', leveraging the which() function and the %in% operator.

In the end, the output is a tibble streamlined to only the genes of interest, with a dedicated column indicating the gene's category.

```{r reduce_data}
# Sample data and gene lists
goodGenes <- c("SLC1A4", "MGAT5B", "LINC02907", "BRIP1", "NRIP1", "TGDS", "PSMD10")
badGenes <- c("MAF", "PLD4", "LIG3", "RARA", "NOTCH4", "USP8", "RARG")

# Applying the function
reduced_data <- reduce_data(result_tib, gene_names, goodGenes, badGenes)
reduced_data

```

## Analysis of Gene Expression Data

In our analysis, we aim to visualize the gene expression data for genes
categorized as 'good' and 'bad'. We are using good and bad arbitrarily to
represent two potentially interesting groups of genes. Before visualizing, it's
crucial to ensure our data is structured correctly. We would like to plot this
data as a boxplot and so we will reformat the data to "long" format to enable
seamless plotting in ggplot2. 

### Convert Data Format
First, we'll look at our reduced data from the previous function. This new
function, `convert_to_long` should properly convert this output to "long" format
with the old sample columns (GSM972389, etc.) in a new column called "sample",
and the values column named 'values' (the default).

```{r convert standard tibble format to long format for plotting}
# original data in "wide" format
reduced_data

# data in long format for easy plotting in ggplot2
long_format_data <- convert_to_long(reduced_data)
long_format_data
```

## Plotting the expression values as a boxplot using ggplot2
Now, with our data in long format, we can visualize the distribution of gene
expression values for the 'good' and 'bad' gene sets. We'll use ggplot2 to
create a quick boxplot:
```{r boxplot from long format data using ggplot2}
library(ggplot2)

long_format_data$hgnc_symbol <- factor(long_format_data$hgnc_symbol,
                            levels = unique(long_format_data$hgnc_symbol), 
                            ordered = TRUE)

ggplot(long_format_data, aes(x = hgnc_symbol, y = value)) + 
  geom_boxplot(aes(fill = gene_set)) +
  labs(title = "These genes were picked randomly, don't expect to see a pattern", 
       x = "Gene Set", 
       y = "Expression Value") +
  theme_minimal() + theme(axis.text.x=element_text(angle = -90, hjust = 0))

```
